### PR TITLE
fix: restore DSA indexes with hybrid HiCache

### PR DESCRIPTION
**Pull request by alphabetc1**
## Motivation

Hybrid Mamba/DSA HiCache restores full KV and recurrent state without restoring `index_k_with_scale_buffer`. After GPU slot reuse, attention can select KV using stale index keys/scales. In a captured GLM-5.3-Flash host replay, the requested `echo hello-repro` became `echo "repro"` in the model's raw output and tool arguments.

Related to #38031. This is a narrow extraction of the index-assembly work from ormandj's #38212; it does not include that PR's separate compressed-prefix ownership, checkpoint, or draft-layout changes.

## Modifications

Register an INDEXER sidecar using the full-KV physical page indices and existing transfer implementation. Map nonempty target/draft index layers, skip empty shared-topk buffers, and account for index storage in the fixed host budget. Mamba transfer kernels are unchanged.

Add one standalone GLM-5.3-Flash NVFP4 host-restoration KL test, following the cache-hit controls in #38474. It requests full-vocabulary output logprobs at eight fixed continuation prefixes and requires substantial host restoration. Every comparison has the same eight conditioning prefixes, even when argmax tokens differ. Scoring must reuse the long prefix and may recompute at most 255 trailing input tokens. Five seeded rounds are mandatory; no implicit CI retry or median gate is used.

The test is registered for `extra-b / 4-gpu-b200` but disabled until the base supports `Glm5NextForConditionalGeneration` and the B200 configuration is calibrated. It can be run directly on a compatible GLM image. Unlike #38474, it computes exact full-vocabulary KL rather than selected-token k3 estimates and gates every per-sequence mean rather than a median. Compressed-prefix branching is outside this test's scope.

### New fixed-prefix KL regression

Completed original/fixed/original/fixed runs on the same 4 B300 GPUs, using the exact committed test bytes (SHA256 `e23f1394a20081a4aaa93efd3070e5866bf74aa00265a461e6e6e5ff7caa10ec`). The gate is **every per-sequence mean KL over eight fixed continuation prefixes `< 0.1`**, including device controls.

| Check | Original | Fixed | Original switchback | Fixed repeat |
|---|---:|---:|---:|---:|
| Maximum host mean KL | **0.150306** | **0.051590** | **0.195408** | **0.060670** |
| Host cases failing the gate | 3/5 | 0/5 | 2/5 | 0/5 |
| Maximum device-control mean KL | 0.041535 | 0.021682 | 0.034721 | 0.075492 |
| Test result | FAIL | PASS | FAIL | PASS |

All 20 device controls passed. All 20 first host scores restored **180032 host tokens plus 128 device tokens**. The following seven scores reuse the restored device-resident prefix. All 40 comparisons contain eight positions; their conditioning continuation prefixes match across all arms. All 420 test generation requests completed. Only the assembler Python file differs between runtime arms. No threshold increase, implicit retry, median gate, or missing-position padding was used.

This run used the pinned image below as a container, Python 3.12.3, PyTorch 2.13.0+cu130, glibc 2.39 and driver 590.48.01. All 49 model-file hashes match the earlier experiment. It is B300 calibration, not a B200 CI pass or universal numerical guarantee. The shared server teardown emits `CancelledError` after requests complete, including in passing runs; logs are not claimed to be pristine.

The registered test is `test/registered/e2e/radix_cache/test_glm53_flash_hicache_restore_kl.py`. To run it on a compatible model implementation:

```bash
SGLANG_TEST_GLM53_MODEL=/models/glm53-nvfp4 \
  python test/registered/e2e/radix_cache/test_glm53_flash_hicache_restore_kl.py
```

### Earlier full tool-call experiment (separate metric/runtime)

Completed a 105-request A/B/A experiment on 4 B300 GPUs with the same model files, inputs and serving configuration:

| Check | Original | Fixed | Original switchback |
|---|---:|---:|---:|
| Correct complete tool arguments | 34/35 | 35/35 | 35/35 |
| Correct host replays | 4/5 | 5/5 | 5/5 |
| Verified host replays | 5/5 | 5/5 | 5/5 |

Every host replay used 180032 host and 128 device tokens. All 90 non-host requests were correct. The error is intermittent; switchback did not reproduce wrong arguments in these five requests, so these samples do not establish a failure-rate reduction or universal correctness.

Asynchronous full-logit captures also show a reversible distribution change. At the same r4 generated prefix (sequence length 180177), full-vocabulary `KL(device-warm || host)` was **0.25731 → 0.00193 → 0.58711**. The host argmax changed from ` simple` to ` bash` in both original runs; the fixed run retained ` simple`. Comparisons use unit-temperature softmax and stop at the first divergent token. Raw argmax tokens match subsequent decode inputs. KL is diagnostic, not a standalone semantic pass/fail gate; clean paths have numerical variation too.

Configuration: ModelScope `RadixArk/GLM-5.3-Flash-NVFP4`, TP4/EP4, DSA/TRTLLM, FP8 KV, speculation off, direct/page_first_direct, write-through, 620000 KV tokens, Mamba cap 128, host size 20 GB, chunked prefill 8192, context 262144, greedy output. Five seeded rounds each flush, cold-replay, device-replay, send four pressure prompts, then host-replay the target. Target filler is 180000 tokens; target seeds start at 919433029 and increase by 100. The failing case is seed 919433429.

Model hashes matched all 38 shards and 11 configuration/tokenizer/template files. Serving configurations differed only in startup timings. Runtime source/packages came from `lmsysorg/sglang@sha256:a2c0f7d4d9ebce97a2707c5415081d284d741db1033a1008a955453b9b5255bf`, executed natively with CPython 3.12.11, PyTorch 2.13.0+cu130 and driver 580.159.03. Both arms use the same parser overlay. The old-image adapter omits its unavailable `page_aligned_only` constructor flag; all tested transfers are whole pages. This is image-overlay validation, not an upstream-main GLM serving claim.

Real index-byte roundtrips also passed all four direct/kernel and layer-first/page-first layout combinations, checking host contents, restored and untouched device bytes, shuffled pages and empty-layer compaction. Pre-commit and commit hooks passed.

## Speed Tests and Profiling

No controlled performance claim.

### CODE REVIEW

**Comment by llc-kc** [comment]
Using DeepSeekV4PagedHostPool feels a bit unrelated. Can we reuse the existing logic like https://github.com/sgl-project/sglang/pull/39698?

**Comment by 1e4ves** [comment]
may be because GLM-5.3-Flash also uses a compressed indexer (index_kpool=4), similar to DSv4’s C4 indexer？@llc-kc

**Comment by llc-kc** [comment]
@1e4ves Thanks, I missed that.

**Comment by llc-kc** [comment]
The device side reuse DSATokenToKVPool IndexKeyCache, and device cache buffer shape is (num_pages, pool.page_size * (pool.index_head_dim + pool.index_head_dim // pool.quant_block_size * 4)).
It’s seems 4 continous token pages reuse one index page kv data? Why do not reduce DSATokenToKVPool IndexKeyCache kv size allocation to (num_pages, 1 / compress_ratio * pool.page_size * (pool.index_head_dim + pool.index_head_dim // pool.quant_block_size * 4)), to make each page independent?

**Comment by llc-kc** [comment]
@alphabetc1 @1e4ves @JustinTong0323  Should we optimize the  DSA indexer compressor kernel so that tokens of each page are stored in their corresponding page, instead of using a scheme where the page size is 64 but 256 tokens share one page? This design makes the HiCache infra extremely complex, hard to understand and maintain.

**Comment by llc-kc** [comment]
Is my understanding correct? If we have 512 tokens in total. The first 256 tokens occupy 4 pages. Although DSA allocates 4 pages, only page 0 is used while the other 3 remain idle. The second 256 tokens also correspond to 4 pages, and DSA stores all of them into page 4. If this is true, a simple solution to avoid modifying HiCache's general implementation and DSA kernels would be to enforce a tree page size of 256, which maintains compatibility with existing logic. However, the current page size is forced to 64 and cannot be modified.
If page size is 64 and two requests prefix branch at length of 64, then they would write different cache to the same indexer kv for length > 64 part. So maybe we should also import tree page size setting part of PR https://github.com/sgl-project/sglang/pull/38212 if we adopt this solution @alphabetc1 
Another solution would be better: since we do not reduce kv cache storage size, we don't need pack 256 token into one page, just simply write the compressed 1/4 indxer cache to first 1/4 part of corresponding indxer page.

**Comment by llc-kc** [comment]
I have implemented a demo that store index cache of each page independently, do you have any suggestions: https://github.com/sgl-project/sglang/pull/40134. If we store index cache of each page independently, we do not need extra logics that set different tree page and server ages tree page and compatible with existing hicache logic.
